homeassistant/components/light/rest.py

```diff
@@ -0,0 +1,432 @@
+                    self._resource)
+        except (asyncio.TimeoutError, aiohttp.ClientError):
+            _LOGGER.error("Error while turn off %s", self._resource)
+
+    @asyncio.coroutine
+    def set_device_state(self, body):
+        """Send a state update to the device."""
+        websession = async_get_clientsession(self.hass)
+
+        with async_timeout.timeout(self._timeout, loop=self.hass.loop):
+            req = yield from getattr(websession, self._method)(
+                self._resource, auth=self._auth, data=bytes(body, 'utf-8'))
+            return req
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Get the current state, catching errors."""
+        try:
+            yield from self.get_device_state(self.hass)
+        except (asyncio.TimeoutError, aiohttp.ClientError):
+            _LOGGER.exception("Error while fetch data.")
+
+    @asyncio.coroutine
+    def get_device_state(self, hass):
+        """Get the latest data from REST API and update the state."""
+        websession = async_get_clientsession(hass)
+
+        with async_timeout.timeout(self._timeout, loop=hass.loop):
+            req = yield from websession.get(self._resource, auth=self._auth)
+            text = yield from req.text()
+
+        result = self._is_on_template.\
+            async_render_with_possible_json_value(text, 'None')
+        result = result.lower()
+        if result == 'true':
+            self._state = True
+        elif result == 'false':
+            self._state = False
+        else:
+            self._state = None
+
+        result = self._supported_features_template.\
+            async_render_with_possible_json_value(text, '')
+        if result != '':
+            self._supported_features = ast.literal_eval(result)
+
+        if "brightness" in self._supported_features:
+            result = self._brightness_template.\
+                async_render_with_possible_json_value(text, 0)
+            self._brightness = int(result)
+
+        if "color_temp" in self._supported_features:
+            result = self._color_temp_template.\
+                async_render_with_possible_json_value(text, 0)
+            self._color_temp = int(result)
+
+        if "effect" in self._supported_features:
+            result = self._effect_template.\
+                async_render_with_possible_json_value(text, '')
+            self._effect = result
+            result = self._effect_list_template.\
+                async_render_with_possible_json_value(text, '[]')
+            self._effect_list = ast.literal_eval(result)
+
+        if "rgb_color" in self._supported_features:
+            result = self._rgb_color_template.\
+                async_render_with_possible_json_value(text, '[]')
+            self._rgb_color = ast.literal_eval(result)
+
+        result = self._name_template.\
+            async_render_with_possible_json_value(text, self._name)
+        self._name = result
+
+        return req
```

Comment:
I don't think you need to return `req` here, as you've `yield`ed from it above.

Reply:
The return value is used in `async_setup_platform`.

Reply:
@tinloaf - my understanding is the same as @MartinHjelmare - I believe I need to return it as  the return value is used in `async_setup_platform`. Please let me know if I've misunderstood.